### PR TITLE
Reboot machine before kexec on UEFI kernel update

### PR DESCRIPTION
**Pull request by dzedro**
- Related ticket: https://progress.opensuse.org/issues/125423
- Verification run: https://openqa.suse.de/tests/overview?build=kexec&distri=sle

### CODE REVIEW

File: lib/serial_terminal.pm [hunk 1]
```diff
@@ -367,6 +367,12 @@ sub select_serial_terminal {
     my $backend = get_required_var('BACKEND');
     my $console;
 
+    if (get_var('OFW') && !check_var('OFW_HVC_SERIAL_CONSOLE_ADDED', '1')) {
+        select_console 'root-console';
+        prepare_serial_console;
+        set_var('OFW_HVC_SERIAL_CONSOLE_ADDED', 1);
+    }
```
Comment (pevik) [inline lib/serial_terminal.pm:374]: I'm ok with this, but wouldn't be better to add the guarder to `prepare_serial_console()`?

https://github.com/os-autoinst/os-autoinst-distri-opensuse/blob/c40110c3ef1ef43138c7547ad0a3ecb50eaf22d7/lib/serial_terminal.pm#L71-L78

BTW adding `prepare_serial_console()` would mean that the other calls below could be deleted, right?
https://github.com/os-autoinst/os-autoinst-distri-opensuse/blob/c40110c3ef1ef43138c7547ad0a3ecb50eaf22d7/tests/console/system_prepare.pm#L33-L38
https://github.com/os-autoinst/os-autoinst-distri-opensuse/blob/c40110c3ef1ef43138c7547ad0a3ecb50eaf22d7/tests/shutdown/cleanup_before_shutdown.pm#L50-L56

**Review by pevik — APPROVED** [review]
Generally LGTM, thanks!

File: lib/qam.pm [hunk 2]
```diff
@@ -30,15 +30,15 @@ sub capture_state {
         my $compression = is_sle('=12-sp1') ? 'bz2' : 'xz';
         assert_script_run "save_y2logs /tmp/y2logs_$state.tar.$compression";
         upload_logs "/tmp/y2logs_$state.tar.$compression";
-        save_screenshot();
+        save_screenshot() unless is_serial_terminal;
```
[the remark this answers is not in the document]
Comment (dzedro) [reply lib/qam.pm:33]: Great, will remove that then. :+1:

File: lib/serial_terminal.pm (hunk 1 above)
Comment (dzedro) [reply lib/serial_terminal.pm:374]: I think only ppc64le does require to start hvc consoles. I will remove other `prepare_serial_console`

Comment (pevik) [reply lib/serial_terminal.pm:374]: I suppose code in `prepare_serial_console()` should be correct (feel free to retest it and suggest change):
* `hvc0` is needed only for SLE < 12-SP2 on all but s390x. NOTE: we still test SLE11-SP4 on kernel (it should be kept)
* `hvc1` is needed for ppc64le. Maybe `get_var('OFW')` could be replaced by `is_ppc64le`, but not sure
* `hvc1` is also needed for other archs except ppc64le for `user-virtio-terminal`
* `hvc2` is needed for ppc64le for `user-virtio-terminal`.
But IMHO 1) call `prepare_serial_console()` only in `prepare_serial_console()` and guard is just by new variable `SERIAL_CONSOLE_ADDED` (it should be for any arch) 2) remove all `prepare_serial_console()` calls elsewhere.

Comment (pevik) [reply lib/serial_terminal.pm:374]: I'd hope that all code uses `select_serial_terminal()` or `select_user_serial_terminal()` so it should work, but maybe I'm missing something. But sure, postpone this change to separate effort.

File: lib/qam.pm (hunk 2 above)
[the remark this answers is not in the document]
Comment (dzedro) [reply lib/qam.pm]: Ok, within the ticket I can clean the rest of unless ...

File: lib/serial_terminal.pm (hunk 1 above)
Comment (pevik) [reply lib/serial_terminal.pm:374]: I've read that before, but that expects all different installer (SLES / Tumbleweed / JeOS, ...) need to schedule `tests/installation/first_boot.pm` (I know they will need it for different things anyway). Also we have preinstalled qcow for old SLES versions (where installer is not tested any more), therefore this setup must be done manually. Having required config in `select_serial_terminal()` would not need it.

Comment (mdoucha) [reply lib/serial_terminal.pm:374]: Or we can move `console/system_prepare.pm` immediately after `installation/*` modules. That'll also make serial terminal available for the rest of post-install setup.

**Review by pevik — APPROVED** [review]

File: tests/qam-minimal/install_update.pm [hunk 3]
```diff
@@ -80,6 +81,13 @@ sub run {
 
         capture_state('between', 1);
 
+        # old kernel does not have key of new kernel
+        if (check_var('MACHINE', 'uefi') && get_var('BUILD') =~ /dtb|kernel/) {
```
Comment (mdoucha) [inline tests/qam-minimal/install_update.pm:85]: Instead of checking `BUILD`, it'd be safer to search incident repos for kernel packages.

**Review by mdoucha — CHANGES_REQUESTED** [review]

File: tests/qam-minimal/install_update.pm (hunk 3 above)
Comment (dzedro) [reply tests/qam-minimal/install_update.pm:85]: It would be better, but there are updates with specific build names.
IMO filtering repo packages is more complicated with same result.

File: lib/serial_terminal.pm (hunk 1 above)
Comment (dzedro) [reply lib/serial_terminal.pm]: `system_prepare` was not scheduled in qam-minimal tests, I will keep it as it is and just schedule `system_prepare`.
If it would be only about hvc console on ppc64 then I would do it within select_serial_terminal, but it is adding also user console, so I think it's good as it is, done in `system_prepare` and after installation in `cleanup_before_shutdown`

File: tests/qam-minimal/install_update.pm (hunk 3 above)
Comment (mdoucha) [reply tests/qam-minimal/install_update.pm:85]: Those build names can change unpredictably on new SLE releases. Searching the repos does not seem complicated to me at all:

    script_run("zypper -n search kernel -r $repo") == 0

Comment (dzedro) [reply tests/qam-minimal/install_update.pm:85]: True, I had complicated idea in head ...
Thanks, will search for kernel in incident repos.

File: tests/qam-minimal/install_patterns.pm [hunk 4]
```diff
@@ -60,7 +61,8 @@ sub run {
     script_run('sed -i -r "s/^DISPLAYMANAGER_AUTOLOGIN/#DISPLAYMANAGER_AUTOLOGIN/" /etc/sysconfig/displaymanager');
     script_run('sed -i -r "s/^DEFAULT_WM=\"icewm\"/DEFAULT_VM=\"\"/" /etc/sysconfig/windowmanager');
     # now we have gnome installed - restore DESKTOP variable
-    set_var('DESKTOP', get_var('FULL_DESKTOP'));
+    set_var('DESKTOP', 'gnome');
+    bmwqemu::save_vars();
```
Comment (mdoucha) [inline tests/qam-minimal/install_patterns.pm:65]: I think this should go to a separate PR.

File: tests/qam-minimal/s390tools.pm [hunk 5]
```diff
@@ -35,7 +36,10 @@ sub run {
     validate_script_output 'lsdasd', sub { m/Bus-ID|Status|Name|Device|Type|BlkSz|Size|Blocks/ };
     validate_script_output 'lsmem', sub { m/RANGE|SIZE|STATE|REMOVABLE|BLOCK/i };
     # dasd_reload does exit with 4
-    assert_script_run 'dasd_reload|grep -E "offline|Activating"';
+    if (script_run('dasd_reload|grep "line 79: echo: write error"') == 0) {
+        record_soft_failure 'bsc#';
+        script_run 'dasd_reload|grep -E "offline|Activating"';
+    }
```
Comment (mdoucha) [inline tests/qam-minimal/s390tools.pm:42]: This should also go to another PR and the BSC number is missing.

File: tests/qam-minimal/install_patterns.pm (hunk 4 above)
Comment (dzedro) [reply tests/qam-minimal/install_patterns.pm:65]: All this things come up when I schedule the test, It will be in separated commit or I can create new PR.
All is WIP, at the end to merge this I have to have all this tests passing on all versions.

File: tests/qam-minimal/s390tools.pm (hunk 5 above)
Comment (dzedro) [reply tests/qam-minimal/s390tools.pm:42]: I know :smile: I just wanted to run it in serial terminal to copy the output. WIP

**Review by foursixnine — APPROVED** [review]
looks good to me overall

File: tests/qam-minimal/install_patterns.pm (hunk 4 above)
Comment (dzedro) [reply tests/qam-minimal/install_patterns.pm]: Done, save_vars is part of set_var when `reload_needles => 1` is used which is also needed for following x11 tests.

File: tests/qam-minimal/install_patterns.pm [hunk 6]
```diff
@@ -60,15 +63,16 @@ sub run {
     script_run('sed -i -r "s/^DISPLAYMANAGER_AUTOLOGIN/#DISPLAYMANAGER_AUTOLOGIN/" /etc/sysconfig/displaymanager');
     script_run('sed -i -r "s/^DEFAULT_WM=\"icewm\"/DEFAULT_VM=\"\"/" /etc/sysconfig/windowmanager');
     # now we have gnome installed - restore DESKTOP variable
-    set_var('DESKTOP', get_var('FULL_DESKTOP'));
+    set_var('DESKTOP', 'gnome', reload_needles => 1);
```
[the remark this answers is not in the document]
Comment (dzedro) [reply tests/qam-minimal/install_patterns.pm:66]: Because the test was designed to start with minimal installation, without X and updates, update system and update RRs.

File: tests/qam-minimal/install_update.pm [hunk 7]
```diff
@@ -80,6 +90,13 @@ sub run {
 
         capture_state('between', 1);
 
+        # old kernel does not have key of new kernel
+        if (check_var('MACHINE', 'uefi') && script_run('zypper se ' . join(' ', map { "-r $_" } split(',', $repo)) . ' kernel') == 0) {
```
[the remark this answers is not in the document]
Comment (dzedro) [reply tests/qam-minimal/install_update.pm]: done

File: tests/qam-minimal/install_patterns.pm (hunk 6 above)
[the remark this answers is not in the document]
Comment (dzedro) [reply tests/qam-minimal/install_patterns.pm:66]: Then run console and X11 tests

**Review by mdoucha — APPROVED** [review]

File: tests/qam-minimal/install_update.pm [hunk 8]
```diff
@@ -19,15 +19,25 @@ use strict;
...
 use utils;
-use power_action_utils qw(prepare_system_shutdown power_action);
+use power_action_utils qw(power_action);
 use version_utils 'is_sle';
 use qam;
 use testapi;
 use Utils::Systemd 'disable_and_stop_service';
+use serial_terminal qw(select_serial_terminal);
+
+sub is_package_in_repo {
```
[the remark this answers is not in the document]
Comment (dzedro) [reply tests/qam-minimal/install_update.pm:29]: Forgot it. Removing

**Review by pevik — APPROVED** [review]
